pkg/k8s/apis/cilium.io/v2alpha1/cec_types.go

```diff
@@ -0,0 +1,162 @@
+	metav1.TypeMeta `json:",inline"`
+	// +k8s:openapi-gen=false
+	// +deepequal-gen=false
+	metav1.ObjectMeta `json:"metadata"`
+
+	// +k8s:openapi-gen=false
+	// +kubebuilder:validation:Type=object
+	Spec CiliumEnvoyConfigSpec `json:"spec,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +deepequal-gen=false
+
+// CiliumEnvoyConfigList is a list of CiliumEnvoyConfig objects.
+type CiliumEnvoyConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	// Items is a list of CiliumEnvoyConfig.
+	Items []CiliumEnvoyConfig `json:"items"`
+}
+
+type CiliumEnvoyConfigSpec struct {
+	// Services specifies Kubernetes services for which traffic is
+	// forwarded to an Envoy listener for L7 load balancing. Backends
+	// of these services are automatically synced to Envoy usign EDS.
+	//
+	// +kubebuilder:validation:Optional
+	Services []*ServiceListener `json:"services,omitempty"`
+
+	// BackendServices specifies Kubernetes services whose backends
+	// are automatically synced to Envoy usign EDS.
+	//
+	// +kubebuilder:validation:Optional
+	BackendServices []*Service `json:"backendServices,omitempty"`
+
+	// Ingress must be set to 'true' when listeners included in
+	// 'resources' are ingress listeners (capturing traffic before
+	// entering pods). Defaults to 'false' for egress listeners
+	// (capturing traffic exiting pods). This implies that all
+	// listeners in 'resources' must be either ingress or egress
+	// listeners.
+	//
+	// This setting directs policy principal selection
+	// (destination for ingress, source for egress) and must be
+	// set properly.
+	//
+	// +kubebuilder:validation:Optional
+	Ingress bool `json:"ingress,omitempty"`
+
+	// Envoy xDS resources
+	// +kubebuilder:validation:Required
+	Resources []XDSResource `json:"resources,omitempty"`
+}
+
+type Service struct {
+	// Name is the name of a destination Kubernetes service that identifies traffic
+	// to be redirected.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// Namespace is the Kubernetes service namespace.
+	// +kubebuilder:validation:Required
+	Namespace string `json:"namespace"`
+}
+
+type ServiceListener struct {
+	// Name is the name of a destination Kubernetes service that identifies traffic
+	// to be redirected.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// Namespace is the Kubernetes service namespace.
+	// +kubebuilder:validation:Required
+	Namespace string `json:"namespace"`
+
+	// Listener specifies the name of the Envoy listener the
+	// service traffic is redirected to. The listener must be
+	// specified in the Envoy 'resources' of the same
+	// CiliumEnvoyConfig.
+	//
+	// If omitted, the first listener specified in 'resources' is
+	// used.
+	//
+	// +kubebuilder:validation:Optional
+	Listener string `json:"listener"`
+}
+
+// +kubebuilder:pruning:PreserveUnknownFields
+type XDSResource struct {
+	*anypb.Any `json:"-"`
```

Comment:
nit: could you add a comment here which xds resource can be specified here? i'm guessing it can be any resource in https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol?

Reply:
Added this comment:
```
// XDSResource can carry the following Envoy resource types:
// - type.googleapis.com/envoy.config.listener.v3.Listener
// - type.googleapis.com/envoy.config.route.v3.RouteConfiguration
// - type.googleapis.com/envoy.config.cluster.v3.Cluster
// - type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment
// - type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret
//
// +kubebuilder:pruning:PreserveUnknownFields
type XDSResource struct {
	*anypb.Any `json:"-"`
}
```

Reply:
I had to move this comment into `CiliumEnvoyConfigSpec.Resources` so that this info would appear in the CRD schema description of that field.